flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,449 @@
+
+### Capabilities are values
+
+Capabilities are a value type. This means that any capability that an account has access to can be copied, and potentially given to someone else. The copied capability will use the link on the same path as the original capability
+
+## Problem statement
+
+There are two potential pain points in the current capability API.
+
+The first one is that capabilities/linking/revoking/redirecting are hard to understand for new developers that are coming to Flow, as that is a lot of new concepts to grasp before one can get started.
+
+The second is that issuing and managing private capabilities that can be revoked at a granular level, by creating a custom private linked path for each capability, is difficult.
+
+- Unless the path name includes some hints, there is no way to know when a path was created, thus making it more difficult to remember why a certain capability was issued.
+- If an old path (that has been unlinked) is accidentally reused and re-linked, this will revive a capability that is supposed to be revoked.
+- Copying a capability is easy (since it is a value type) but doing so might give unintended access to third parties.
+
+## Suggested change
+
+The suggested change addresses these pain points by:
+
+- Removing (or abstracting away) the concept of links.
+- Making it easier to create new capabilities (with individual links).
+- Offering a way to iterate through capabilities on a storage path.
+- Removing the need to have a `/private/`domain.
+- Introducing Capability Controllers (CapCons) that handle management of capabilities.
+
+### Capability Controllers (CapCons)
+
+Each Capability would have an associated CapCon that would be created together with the Capability (Capabilities and Capability Controllers are in a 1 to 1 relation). The data associated with CapCons would be stored in arrays, so that each storage path on an account has an array of CapCons of Capabilities issued from that storage path.
+
+CapCons are a non-storable object.
+
+Capabilities also have an address field and a target field. The target field is the storage path of the targeted object. Creating a capability from a capability should be illegal.
+
+The definition of the `CapabilityController` .
+
+```cadence
+// CapabilityController can be retrieved via:
+// - AuthAccount::getControllers(path: StoragePath): [CapabilityController]
+// - AuthAccount::getController(capabilityID: UInt64): CapabilityController?
+struct CapabilityController {
+   // The block height when the capability was created.
+   let issueHeight: UInt64
+   // The Type of the capability.
+   let CapabilityType: Type
+   // The id of the related capability.
+   // This is the UUID of the created capability.
+   let capabilityID: UInt64
+
+   // Is the capability revoked.
+   fun isRevoked(): Bool
+   // The target storage path.
+   fun target(): StoragePath
+   // Revoke the capability.
+   fun revoke()
```

Comment:
It would be great if the proposal could describe the behaviour of revokation in more detail